common/src/heuristics.rs

```diff
@@ -8,53 +12,241 @@ const SIGSEGV: u32 = 11;
+/// elides its own code from the return address.
+///
+/// # Heuristics
+///
+/// To resolve this particular issue, it is necessary for the symbolication implementor to apply a
+/// per-architecture heuristics to the return addresses, and thus derive the **likely** address of
+/// the actual calling instruction. There is a high probability of correctness, but absolutely no
+/// guarantee.
+///
+/// This derived address **should** be used as the symbolication address, but **should not** replace
+/// the return address in the crash report. This derived address is a best guess, and if you replace
+/// the return address in the report, the end-user will have lost access to the original canonical
+/// data from which they could have made their own assessment.
+///
+/// These heuristics must not be applied to frame #0 on any thread. The first frame of all threads
+/// contains the actual register state of that thread at the time that it crashed (if it's the
+/// crashing thread), or at the time it was suspended (if it is a non-crashing thread). These
+/// heuristics should only be applied to frames *after* frame #0 – that is, starting with frame #1.
+///
+/// Additionally, these heuristics assume that your symbolication implementation correctly handles
+/// addresses that occur within an instruction, rather than directly at the start of a valid
+/// instruction. This should be the case for any reasonable implementation, but is something to be
+/// aware of when deploying these changes.
+///
+/// ## x86 and x86-64
+///
+/// x86 uses variable-width instruction encodings; subtract one byte from the return address to
+/// derive an address that should be within the calling instruction. This will provide an address
+/// within a calling instruction found directly prior to the return address.
+///
+/// ## ARMv6 and ARMv7
 ///
-/// See <https://goo.gl/g17EAn> for detailed information on this topic.
+/// - **Step 1:** Strip the low order thumb bit from the return address. ARM uses the low bit to
+///   inform the processor that it should enter thumb mode when jumping to the return address. Since
+///   all instructions are at least 2 byte aligned, an actual instruction address will never have
+///   the low bit set.
+///
+/// - **Step 2:** Subtract 2 Bytes. 32-bit ARM instructions are either 2 or 4 bytes long, depending
+///   on the use of thumb. This will place the symbolication address within the likely calling
+///   instruction. All ARM64 instructions are 4 bytes long; subtract 4 bytes from the return address
+///   to derive the likely address of the calling instruction.
+///
+/// # More Information
+///
+/// The above information was taken and slightly updated from the now-gone *PLCrashReporter Wiki*.
+/// An old copy can still be found in the [internet archive].
+///
+/// [internet archive]:
+/// https://web.archive.org/web/20161012225323/https://opensource.plausible.coop/wiki/display/PLCR/Automated+Crash+Report+Analysis
```

Comment:
does this need to be on one line rather?

```suggestion
/// [internet archive]: https://web.archive.org/web/20161012225323/https://opensource.plausible.coop/wiki/display/PLCR/Automated+Crash+Report+Analysis
```

Reply:
Argh, my auto formatter constantly destroys these links.